### PR TITLE
Add CONTRIBUTING doc

### PR DESCRIPTION
**Pull request by iHiD**
**This is WIP and will change lots so don't worry about reviewing it yet. If you have any general comments about things it should include, please add those below though.**

---

This PR covers the addition of a CONTRIBUTING.md document that outlines the technical philosophy and the rules/standards we want to develop the codebase against. I intend to write down everything I feel firmly about as a first pass, and then we can discuss any bits that are unclear etc. Then as the next couple of months evolve we can flesh out the other bits that are maybe less clear at this stage etc. I want this to be a living document that evolves throughout the process, and that by the time of launch is quite firm and immutable, which then makes accepting any contributions from others an easier thing.

### CODE REVIEW

File: docs/CONTRIBUTING.md [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+In practice that means that Commands are responsible for completing some sort of action that might touch different areas of the project.
+For example, creating an iteration (`Iteration::Create.()`), creates multiple db records, writes to DynamoDB and S3, generates notifications and more.
+
+Each Command should have responsibility for doing one domain action (e.g. creating or updating something).
+It should then proxy other parts of that to other Commands (e.g. `Interation::Create` calls `Notification::Create`) or create records specifically under its ownership (e.g. `Iteration::Create.()` calling `Iteration.create()`.
+
+Commands should be highly readable.
+It should be extremely clear from reading the Command's `call` method what the Command does.
+
+Commands should use Mandate.
```
Comment (ErikSchierboom) [inline docs/CONTRIBUTING.md]: Maybe link to the repo?

```suggestion
Commands should use [Mandate](https://github.com/iHiD/mandate).
```

**Review by ErikSchierboom — APPROVED** [review]
Really nice!

File: docs/CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+To ensure that data is represented in common ways, we use serializers.
+For example, the data for the mentoring workspace would be represented via a serializer.
+The HTML that renders the table as a React Component would use the serializer to generate the JSON for the initial React Component, and the API endpoint would use the same Serializer when the table's data is filtered.
+
+## Models
+
+Models are light-weight database wrappers.
+We treat models as highly-predictable, meaning that creating and updating models should not have side-effects.
+We therefore use `before_xxx` and `after-xxx` blocks sparsely.
```
Comment (dougal) [inline docs/CONTRIBUTING.md]: s/sparsely/sparingly

File: docs/CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+The only time that changing a model should change other data is if the model does not make sense without that other data.
+For example, users should **always** have authentication tokens.
+So `User.create` can resonably call `AuthToken.create` in its `after_create` block.
+However, although submitting iterations should create notifications, because it is not essential for the existance of the iteration to make sense, `Iteration.create` would **not** be responsible for calling `Notification.create`.
+
+Model's methods should not cause any side-effects that cannot be infered from the method name.
```
Comment (dougal) [inline docs/CONTRIBUTING.md]: s/infered/inferred

File: docs/CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+- Each component has a method in `app/helpers/components_helper.rb`
+- Each component has a view file for development usage and system testing in `/app/views/test/components/**/XXX.html.haml`
+- Each component has Rails system test in `test/system/components/**/XXX.rb`
+- Each component has Jest test in `test/system/javascript/**/XXX.rb`
+
+The tests should cover all functionality in the component, with unit tests being via JS, and tests that interact with Rails being tested through system tests.
+
+## Linting
+
+Pre-commit linting should happen automatically via our Husky hook.
```
Comment (dougal) [inline docs/CONTRIBUTING.md]: Link the Husky config, or repo.

File: docs/todo.md [hunk 5]
```diff
@@ -1,9 +1,19 @@
-# Personal TODO notes
+# Pre-launchs TODO notes
```
Comment (dougal) [inline docs/todo.md]: s/launchs/launch

File: docs/CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+It is used by other components in our architecture to pass messages and data around.
+It is not accessible via the Internet.
+
+All other "normal" routes redirect or render HTML.
+
+## Controllers
+
+Controllers should be "thin".
+Each action should do at most one thing, retrieve at most one thing, and render/redirect.
+To achieve this, controllers call out to Commands, which contain more complex functionality encapsualted in stand-alone procedures. This is know as the Command Pattern or the Interactor Pattern.
```
Comment (kntsoriano) [inline docs/CONTRIBUTING.md]:
```suggestion
To achieve this, controllers call out to Commands, which contain more complex functionality encapsualted in stand-alone procedures. This is known as the Command Pattern or the Interactor Pattern.
```

File: docs/CONTRIBUTING.md (hunk 2 above)
Comment (kntsoriano) [inline docs/CONTRIBUTING.md]:
```suggestion
We therefore use `before_xxx` and `after_xxx` blocks sparsely.
```

**Comment by kntsoriano** [comment]
Should we add a section about testing? Like having one system test for each user interaction, naming the system test correctly, etc.

File: docs/CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+
+**Please note: We are not currently accepting Pull Requests to this repository.**
+
+## Development Philosophy
+
+This codebase aims to achieve the following goals:
+
+- Optimise for the principle of least surprise for developers
+- Aim to reduce churn and reduce the amount of different files touched during a change.
+- Reduce the likelyhood of bugs.
```
Comment (kntsoriano) [inline docs/CONTRIBUTING.md]:
```suggestion
- Reduce the likelihood of bugs.
```

File: docs/CONTRIBUTING.md (hunk 2 above)
Comment (ErikSchierboom) [reply docs/CONTRIBUTING.md]: @dougal Did you know that GitHub allows for comments to include suggestions? See https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request This makes it really easy for the PR submitter to apply changes. 

For an example what this looks like, see the suggestions made by kntsoriano on this PR.

Comment (dougal) [reply docs/CONTRIBUTING.md]: Ooh, fancy! TIL

Thanks @ErikSchierboom.

Comment (ErikSchierboom) [reply docs/CONTRIBUTING.md]: My pleasure! I've made extensive use of that feature and I love it :D

File: docs/CONTRIBUTING.md (hunk 7 above)
Comment (iHiD) [reply docs/CONTRIBUTING.md]: @kntsoriano I never knew this!!

File: docs/CONTRIBUTING.md (hunk 2 above)
Comment (iHiD) [inline docs/CONTRIBUTING.md]:
```suggestion
We therefore use `before_xxx` and `after-xxx` blocks sparingly.
```

File: docs/CONTRIBUTING.md (hunk 3 above)
Comment (iHiD) [reply docs/CONTRIBUTING.md]:
```suggestion
Model's methods should not cause any side-effects that cannot be inferred from the method name.
```

File: docs/todo.md (hunk 5 above)
Comment (iHiD) [reply docs/todo.md]:
```suggestion
# Pre-launch TODO notes
```

File: docs/CONTRIBUTING.md (hunk 4 above)
Comment (iHiD) [reply docs/CONTRIBUTING.md]:
```suggestion
Pre-commit linting should happen automatically via our [Husky hook](https://github.com/typicode/husky#husky).
```

**Comment by iHiD** [comment]
I'm merging this so that we can start other PRs to this file for all the various things that we're picking up. Thanks for the comments so far.

**Pull request merged**